backend/manager/modules/bll/src/main/java/org/ovirt/engine/core/bll/scheduling/utils/DedicatedCpuPinningHelper.java

```diff
@@ -0,0 +1,212 @@
+package org.ovirt.engine.core.bll.scheduling.utils;
+
+import java.util.ArrayList;
+import java.util.Iterator;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+import org.ovirt.engine.core.common.businessentities.CpuPinningPolicy;
+import org.ovirt.engine.core.common.businessentities.VDS;
+import org.ovirt.engine.core.common.businessentities.VM;
+import org.ovirt.engine.core.common.businessentities.VdsCpuUnit;
+import org.ovirt.engine.core.common.utils.CpuPinningHelper;
+import org.ovirt.engine.core.compat.Guid;
+import org.ovirt.engine.core.vdsbroker.ResourceManager;
+
+import javax.inject.Inject;
+
+public class DedicatedCpuPinningHelper {
+    @Inject
+    private ResourceManager resourceManager;
+
+    /**
+     * This function will tell if the host is capable to run a given dedicated CPU policy VM.
+     *
+     * The function first will pin the CPU pending resources to the host, then:
+     * 1. If the VM number of vCPUs is smaller or equal to the pCPU exists on a single socket:
+     *    Pass on each socket and check if there are enough free pCPU to use.
+     *    If there are enough pCPUs - return true, otherwise false.
+     * 2. If the VM number of vCPUs is bigger to the pCPU exists on a single socket:
+     *    - We must use multiple sockets.
+     *    Pass on the host and check if there are enough free pCPU to use (regardless being in multiple sockets).
+     *    If there are enough pCPUs - return true, otherwise false.
+     *
+     * @param vmToPendingPinnings Map of VDS GUID keys to list of VdsCpuUnits pending to be taken.
+     * @param vm VM object.
+     * @param host VDS object.
+     * @return boolean. True if possible to dedicate the VM on the host. Otherwise false.
+     */
+    public synchronized boolean isDedicatedCpuPinningPossibleAtHost(Map<Guid, List<VdsCpuUnit>> vmToPendingPinnings,
+            VM vm, VDS host) {
+        List<VdsCpuUnit> cpuTopology = clone(resourceManager.getVdsManager(host.getId()).getCpuTopology());
+
+        if (vm.getCpuPinningPolicy() != CpuPinningPolicy.DEDICATED) {
+            // TODO: Implementation for siblings and isolate threads
+            return false;
+        }
+
+        pinPendingExclusiveCpus(cpuTopology, vmToPendingPinnings, vm.getCpuPinningPolicy());
+
+        int cpus = vm.getNumOfCpus();
+        int cpusLeft = cpus;
+        int hostCoresPerSocket = host.getCpuCores() / host.getCpuSockets();
+        int hostThreadsPerCore = host.getCpuThreads() / host.getCpuCores();
+
+        for (int socket = 0; socket < host.getCpuSockets(); socket++) {
```

Comment:
Wouldn't it actually be enough to check `getFreeCpusInSocket(cpuTopology, socket).size() >= cpus` to see if the dedicated policy is possible on the host?

Reply:
Yes, but it will be enough only when the vm fits single socket. I simplified it based on your suggestion.